pkg/image/importer/client.go

```diff
@@ -0,0 +1,377 @@
+			}
+			dockerImage.ID = d.String()
+		} else {
+			d, err := digest.FromBytes(manifest.Raw)
+			if err != nil {
+				return nil, fmt.Errorf("unable to create digest from image bytes: %v", err)
+			}
+			dockerImage.ID = d.String()
+		}
+	}
+	image := &api.Image{
+		ObjectMeta: kapi.ObjectMeta{
+			Name: dockerImage.ID,
+		},
+		DockerImageMetadata:        *dockerImage,
+		DockerImageManifest:        string(manifest.Raw),
+		DockerImageMetadataVersion: "1.0",
+	}
+
+	return image, nil
+}
+
+func schema0ToImage(dockerImage *dockerregistry.Image, id string) (*api.Image, error) {
+	var baseImage api.DockerImage
+	if err := kapi.Scheme.Convert(&dockerImage.Image, &baseImage); err != nil {
+		return nil, fmt.Errorf("could not convert image: %#v", err)
+	}
+
+	image := &api.Image{
+		ObjectMeta: kapi.ObjectMeta{
+			Name: dockerImage.ID,
+		},
+		DockerImageMetadata:        baseImage,
+		DockerImageMetadataVersion: "1.0",
+	}
+
+	return image, nil
+}
+
+func unmarshalDockerImage(body []byte) (*api.DockerImage, error) {
+	var image dockerpre012.DockerImage
+	if err := json.Unmarshal(body, &image); err != nil {
+		return nil, err
+	}
+	dockerImage := &api.DockerImage{}
+	if err := kapi.Scheme.Convert(&image, dockerImage); err != nil {
+		return nil, err
+	}
+	return dockerImage, nil
+}
+
+func isDockerError(err error, code errcode.ErrorCode) bool {
+	switch t := err.(type) {
+	case errcode.Errors:
+		for _, err := range t {
+			if isDockerError(err, code) {
+				return true
+			}
+		}
+	case errcode.ErrorCode:
+		if code == t {
+			return true
+		}
+	case errcode.Error:
+		if t.ErrorCode() == code {
+			return true
+		}
+	}
+	return false
+}
+
+var nowFn = time.Now
+
+type retryRepository struct {
+	distribution.Repository
+
+	retries int
+	initial *time.Time
+	wait    time.Duration
+	limit   time.Duration
+}
+
+// NewRetryRepository wraps a distribution.Repository with helpers that will retry authentication failures
+// over a limited time window and duration. This primarily avoids a DockerHub issue where public images
+// unexpectedly return a 401 error due to the JWT token created by the hub being created at the same second,
+// but another server being in the previous second.
+func NewRetryRepository(repo distribution.Repository, retries int, interval time.Duration) distribution.Repository {
+	var wait time.Duration
+	if retries > 1 {
+		wait = interval / time.Duration(retries-1)
+	}
+	return &retryRepository{
+		Repository: repo,
+
+		retries: retries,
+		wait:    wait,
+		limit:   interval,
+	}
+}
+
+// shouldRetry returns true if the error is not an unauthorized error, if there are no retries left, or if
+// we have already retried once and it has been longer than r.limit since we retried the first time.
+func (r *retryRepository) shouldRetry(err error) bool {
+	if err == nil {
+		return false
+	}
+	if !isDockerError(err, errcode.ErrorCodeUnauthorized) {
+		return false
+	}
+
+	if r.retries <= 0 {
```

Comment:
No concurrent calls (underlying docker client is not thread safe)

On Apr 19, 2016, at 10:54 PM, Jordan Liggitt <notifications@github.com>
wrote:

In pkg/image/importer/client.go
<https://github.com/openshift/origin/pull/8558#discussion_r60344323>:

> +		wait:    wait,
> +		limit:   interval,
> +	}
> +}
> +
> +// shouldRetry returns true if the error is not an unauthorized error, if there are no retries left, or if
> +// we have already retried once and it has been longer than r.limit since we retried the first time.
> +func (r *retryRepository) shouldRetry(err error) bool {
> +	if err == nil {
> +		return false
> +	}
> +	if !isDockerError(err, errcode.ErrorCodeUnauthorized) {
> +		return false
> +	}
> +
> +	if r.retries <= 0 {

don't know the calling code well enough to know if we need to take a lock
here to avoid racing between goroutines

—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/8558/files/9dd0f3197b9856bd93d606507abddfe896f110f3#r60344323>